File: red/src/diagnostics.ts
/** Redacted, bounded diagnostics. Never includes command output or state contents. */
import {accessSync,constants,statSync} from 'node:fs';
import {delimiter,isAbsolute,resolve} from 'node:path';
type Map=Record<string,any>;
export const messages:Record<string,string>={command_failed:'Required command failed.',missing_credentials:'Required credentials are not set.',state_unreadable:'Compute state could not be read.',state_absent:'Required compute state is absent.',identity_mismatch:'Compute state identity does not match the requested node.',unsafe_plan:'Compute plan requires an unauthorized change.',invalid_request:'Invalid compute request.',key_access_failed:'SSH key access could not be prepared.',filesystem_error:'Compute working files could not be accessed.',internal_error:'Compute operation failed.'};
export function redact(stderr:unknown,opts:Map,environment:Map):string|undefined{
 if(typeof stderr!=='string'||!stderr.trim())return undefined;
 let text=stderr.replace(/\x1b\][\s\S]*?(?:\x07|\x1b\\)/g,'').replace(/\x1b\[[0-?]*[ -/]*[@-~]/g,'').replace(/[\p{Cc}\p{Cf}]/gu,char=>char==='\n'||char==='\t'?char:'');
 // Remove blocks first: replacing a known header must never strand a key payload.
 if(/(?:^|\n)\s*[\[{]/.test(text)||/[{]\s*(?:["':]|[A-Za-z][\w-]*\s*:)|\[\s*["'{]/.test(text)||/"(?:resources|planned_values|resource_changes|outputs|private_key_openssh)"\s*:/.test(text))return '[structured output suppressed]';
 text=text.replace(/-----BEGIN [^-\r\n]+-----[\s\S]*?(?:-----END [^-\r\n]+-----|$)/g,'[private material suppressed]');
 const secrets:string[]=[];
 function collect(v:any,key='',inherited=false){const sensitive=inherited||/(?:secret|token|password|credential|authorization|access.?key|api.?key|private.?key)/i.test(key)||key.startsWith('COLORS_PAR_');if(typeof v==='string'&&v&&sensitive)secrets.push(v);else if(v&&typeof v==='object')for(const [k,x] of Object.entries(v))collect(x,k,sensitive);}
 collect(environment);collect(opts);
 for(const secret of secrets.sort((a,b)=>b.length-a.length)){const encoded=JSON.stringify(secret).slice(1,-1);let uri='';try{uri=encodeURIComponent(secret);}catch{}for(const value of new Set([secret,encoded,uri,uri.replaceAll('%20','+'),Buffer.from(secret).toString('base64')]))if(value)text=text.split(value).join('[REDACTED]');}
 text=text.replace(/(authorization\s*[:=]\s*)[^\r\n]*/gi,'$1[REDACTED]').replace(/\bBearer\s+[^\s"'<>]+/gi,'Bearer [REDACTED]');
 text=text.replace(/((?:[A-Za-z0-9_-]*(?:secret|token|password|access[_-]?key|api[_-]?key|private[_-]?key)[A-Za-z0-9_-]*)\s*[:=]\s*)[^\r\n]*/gi,'$1[REDACTED]');
 return text.slice(0,2000);
}
export function executable(program:string,cwd:string,environment:Map):string|undefined{
 const candidates=program.includes('/')?[isAbsolute(program)?program:resolve(cwd,program)]:typeof environment.PATH==='string'?environment.PATH.split(delimiter).map((p:string)=>resolve(cwd,p||'.',program)):[];
 for(const path of candidates)try{if(statSync(path).isFile()){accessSync(path,constants.X_OK);return path;}}catch{}return undefined;
}
export class CommandFailure extends Error {constructor(public details:Map){super('command failed');}}
export function commandFailure(args:string[],cwd:string,env:Map,result:Map,opts:Map,source:Map){const command=args[0]==='tofu'?args.slice(0,args[1]==='state'?3:2):args[0]==='aws'?args.slice(0,3):args[0]==='ssh-keygen'?['ssh-keygen']:args[0]==='gcloud'?args.slice(0,3):[args[0]];const resolved=executable(args[0],cwd,env),stderr=redact(result.err,opts,source);return new CommandFailure({command,...(resolved&&redact(resolved,opts,source)===resolved&&!/[\p{Cc}\p{Cf}]/u.test(resolved)?{executable:resolved}:{}),...(Number.isInteger(result.exit)?{exit_code:result.exit}:{}),...(stderr?{stderr}:{})});}
export function failure(error:unknown,stage:string,infrastructure_changes:string){
 const message=error instanceof Error?error.message:'';let code=error instanceof CommandFailure?(stage==='state'?'state_unreadable':stage==='access'?'key_access_failed':'command_failed'):stage==='credentials'?'missing_credentials':stage==='state'?'state_unreadable':stage==='plan-validation'?'unsafe_plan':stage==='access'?'key_access_failed':['validate','build'].includes(stage)?'invalid_request':'internal_error';
 if(!(error instanceof CommandFailure)&&stage!=='validate'){if(/identity|provider mismatch|provider cannot change|backend cannot change/.test(message))code='identity_mismatch';else if(/absent|state required|state is required|required state/.test(message))code='state_absent';else if(/credential/.test(message))code='missing_credentials';else if(/unsafe.*(?:file|directory)|compute file|OpenTofu file|OpenTofu directory/.test(message)||(error as Map)?.code&&/^(?:EACCES|EPERM|ENOENT|ELOOP|EISDIR|ENOTDIR|EMFILE|ENOSPC)$/.test((error as Map).code))code='filesystem_error';}
 return {status:'error',error:{code,stage,message:messages[code],infrastructure_changes,...(error instanceof CommandFailure?error.details:{})}};
}

File: blue/src/colors_compute/diagnostics.py
"""Authored lifecycle errors and conservative, bounded command diagnostics."""
import base64
from collections.abc import Mapping, Sequence
import json
import os
from pathlib import Path
import re
import unicodedata
from urllib.parse import quote, quote_plus

MESSAGES = {
    'command_failed': 'Required command failed.',
    'missing_credentials': 'Required credentials are not set.',
    'state_unreadable': 'Compute state could not be read.',
    'state_absent': 'Required compute state is absent.',
    'identity_mismatch': 'Compute state identity does not match the requested node.',
    'unsafe_plan': 'Compute plan requires an unauthorized change.',
    'invalid_request': 'Invalid compute request.',
    'key_access_failed': 'SSH key access could not be prepared.',
    'filesystem_error': 'Compute working files could not be accessed.',
    'internal_error': 'Compute operation failed.',
}


class NodeError(Exception):
    def __init__(self, code, **metadata):
        super().__init__(MESSAGES[code])
        self.code = code
        self.metadata = metadata


def credential_values(*sources):
    values = set()
    def visit(value, secret=False):
        if isinstance(value, Mapping):
            for key, item in value.items():
                visit(item, secret or bool(re.search(r'token|secret|password|credential|private.?key|access.?key|api.?key|authorization', str(key), re.I)))
        elif isinstance(value, Sequence) and not isinstance(value, (str, bytes, bytearray)):
            for item in value:
                visit(item, secret)
        elif secret and isinstance(value, str) and value:
            values.add(value)
    for source in sources:
        visit(source)
    return values


def sanitize_stderr(value, secrets):
    if not isinstance(value, str):
        return ''
    value = re.sub(r'\x1b\][^\x07]*(?:\x07|\x1b\\)', '', value)
    value = re.sub(r'\x1b\[[0-?]*[ -/]*[@-~]', '', value)
    value = ''.join(char for char in value if char in '\n\t' or unicodedata.category(char) not in ('Cc', 'Cf'))
    value = re.sub(r'-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----[\s\S]*?(?:-----END [A-Z0-9 ]*PRIVATE KEY-----|$)', '[private material suppressed]', value)
    for secret in sorted(secrets, key=len, reverse=True):
        variants = {secret, json.dumps(secret)[1:-1], json.dumps(secret, ensure_ascii=False)[1:-1], quote(secret, safe=''), quote_plus(secret, safe=''), base64.b64encode(secret.encode()).decode(), base64.urlsafe_b64encode(secret.encode()).decode()}
        variants.update(re.sub(r'%[0-9A-F]{2}', lambda match: match[0].lower(), value) for value in tuple(variants))
        for variant in sorted(variants, key=len, reverse=True):
            if variant:
                value = value.replace(variant, '[REDACTED]')
    value = re.sub(r"(?im)(\b(?:[A-Za-z0-9_-]*(?:token|secret|password|private[_-]?key|access[_-]?key|api[_-]?key)[A-Za-z0-9_-]*|authorization)\b[\"']?\s*[=:]\s*)[^\n]*", r'\1[REDACTED]', value)
    value = re.sub(r'(?i)\bBearer\s+\S+', 'Bearer [REDACTED]', value)
    if re.search(r'(?:\{\s*[:"\'}]|\[\s*["\'{\[]|"(?:resources|outputs|planned_values|resource_changes|private_key)"\s*:)', value):
        return '[structured output suppressed]'
    return value[:2000]


def command_metadata(args, environment, directory, result=None):
    executable = str(args[0])
    name = Path(executable).name
    if name == 'tofu':
        command = [name, *args[1:3]] if args[1:2] == ['state'] else [name, *args[1:2]]
    elif name in ('aws', 'gcloud'):
        command = [name, *args[1:3]]
    elif name == 'ssh-keygen':
        command = [name]
    else:
        command = [name]
    candidates = ([Path(directory) / executable] if '/' in executable else
                  [(Path(part) if Path(part).is_absolute() else Path(directory) / part) / executable
                   for part in (environment['PATH'].split(os.pathsep) if 'PATH' in environment else [])])
    resolved = next((os.path.abspath(path) for path in candidates if path.is_file() and os.access(path, os.X_OK)), None)
    metadata = {'command': command}
    if resolved:
        metadata['executable'] = resolved
    if result is not None:
        if type(result.exit) is int:
            metadata['exit_code'] = result.exit
        metadata['stderr'] = result.err
    return metadata


def classify(error, stage):
    if isinstance(error, NodeError):
        return error.code
    if isinstance(error, OSError):
        return 'filesystem_error'
    message = str(error)
    if isinstance(error, ValueError):
        if any(word in message for word in ('identity mismatch', 'identity changed', 'provider change', 'provider or unit change', 'remote key ownership', 'backend identity')):
            return 'identity_mismatch'
        if any(word in message for word in ('state missing', 'state is missing', 'state is absent', 'existing state required', 'state required')):
            return 'state_absent'
        if 'unsafe' in message and any(word in message for word in ('file', 'directory')):
            return 'filesystem_error'
    return {'validate': 'invalid_request', 'build': 'filesystem_error', 'credentials': 'missing_credentials',
            'state': 'state_unreadable', 'plan-validation': 'unsafe_plan', 'access': 'key_access_failed',
            'cleanup': 'filesystem_error'}.get(stage, 'internal_error')


def failure(error, stage, changes, secrets):
    code = classify(error, stage)
    detail = {'code': code, 'stage': stage, 'message': MESSAGES[code], 'infrastructure_changes': changes}
    if isinstance(error, NodeError):
        detail.update(error.metadata)
        if 'stderr' in detail:
            detail['stderr'] = sanitize_stderr(detail['stderr'], secrets)
        if 'executable' in detail and (any(secret in detail['executable'] for secret in secrets) or any(unicodedata.category(char) in ('Cc', 'Cf') for char in detail['executable'])):
            detail.pop('executable')
    return {'status': 'error', 'error': detail}
